### PR TITLE
Dan/refactor terraform shell

### PR DESCRIPTION
**Pull request by qualidan**

### CODE REVIEW

File: package/cloudshell/iac/terraform/services/local_dir_service.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import os
+import shutil
+from pathlib import Path
+
+from cloudshell.iac.terraform.downloaders.downloader import Downloader
+
+
+class LocalDir:
+    @staticmethod
+    def delete_local_temp_dir(sandbox_data_handler, tf_working_dir):
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/services/local_dir_service.py]: please add typing

File: package/cloudshell/iac/terraform/services/local_dir_service.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        tf_path_str = str(tf_path)
+        shutil.rmtree(tf_path_str)
+        sandbox_data_handler.set_tf_working_dir("")
+
+    @staticmethod
+    def does_working_dir_exists(dir: str) -> bool:
+        return dir and os.path.isdir(dir)
+
+    @staticmethod
+    def prepare_tf_working_dir(logger, sandbox_data_handler, shell_helper):
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/services/local_dir_service.py]: please add typing

File: package/cloudshell/iac/terraform/services/object_factory.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+from cloudshell.iac.terraform.services.svc_attribute_handler import ServiceAttrHandler
+from cloudshell.iac.terraform.services.tf_proc_exec import TfProcExec
+from cloudshell.iac.terraform.tagging.tags import TagsManager
+
+
+class ObjectFactory:
+
+    @staticmethod
+    def create_tf_proc_executer(config: TerraformShellConfig,
+                                sandbox_data_handler, shell_helper, tf_working_dir):
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/services/object_factory.py]: please add typing

File: package/cloudshell/iac/terraform/services/sandbox_messages.py [hunk 4]
```diff
@@ -17,3 +17,6 @@ def _format_message(self, service_name, message):
         if ' ' in service_name:
             service_name = '"{0}"'.format(service_name)
         return '{0} {1}'.format(service_name, message)
+
+    def write_error_message(self, message: str):
+        self.write_message(f'<p style="color:red">{message}</p>')
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/services/sandbox_messages.py]: Did you actual tested that it looks good in CloudShell using the UI?

File: package/cloudshell/iac/terraform/terraform_shell.py [hunk 5]
```diff
@@ -1,149 +1,98 @@
...
-
-    def _create_tf_proc_executer(self, sandbox_data_handler, shell_helper, tf_working_dir):
-        backend_handler = BackendHandler(shell_helper, tf_working_dir, sandbox_data_handler.get_tf_uuid())
-        input_output_service = InputOutputService(shell_helper, self._config.inputs_map, self._config.outputs_map)
-        tf_proc_executer = TfProcExec(shell_helper, sandbox_data_handler, backend_handler, input_output_service)
-        return tf_proc_executer
\ No newline at end of file
+    @staticmethod
+    def _does_working_dir_exists(directory: str) -> bool:
+        return directory and os.path.isdir(directory)
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/terraform_shell.py]: maybe this should be moved to the LocalDir service?

**Review by alexazarh — CHANGES_REQUESTED** [review]

**Comment by alexazarh** [comment]
@qualidan please see my comments

File: package/cloudshell/iac/terraform/services/sandbox_messages.py [hunk 6]
```diff
@@ -8,15 +8,17 @@ def __init__(self, api: CloudShellAPISession, sandbox_id: str, service_name: str
...
 
     def _format_message(self, service_name, message):
         if ' ' in service_name:
             service_name = '"{0}"'.format(service_name)
         return '{0} {1}'.format(service_name, message)
 
     def write_error_message(self, message: str):
-        self.write_message(f'<p style="color:red">{message}</p>')
+        prefix = f'<span style="text-transform:lowercase"><p style="display:inline"><font color="red">'
+        postfix = '</p></span>'
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/services/sandbox_messages.py]: in line 22 string formatting is not needed
in line 23 you dont close the "font" tag

File: package/cloudshell/iac/terraform/terraform_shell.py [hunk 7]
```diff
@@ -1,98 +1,149 @@
...
+        return TerraformServiceObject(api, reservation_id, name, cloudshell_model_name)
+
+    def _does_working_dir_exists(self, dir: str) -> bool:
+        return dir and os.path.isdir(dir)
+
+    def _create_tf_proc_executer(self, sandbox_data_handler, shell_helper, tf_working_dir):
+        backend_handler = BackendHandler(shell_helper, tf_working_dir, sandbox_data_handler.get_tf_uuid())
+        input_output_service = InputOutputService(shell_helper, self._config.inputs_map, self._config.outputs_map)
+        tf_proc_executer = TfProcExec(shell_helper, sandbox_data_handler, backend_handler, input_output_service)
+        return tf_proc_executer
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/terraform_shell.py]: in a previous commit it was in a factory, why did you move it back?

**Review by alexazarh — CHANGES_REQUESTED** [review]
@qualidan why did you revert the changes with the factory class?

File: package/cloudshell/iac/terraform/services/sandbox_messages.py (hunk 6 above)
Comment (qualidan) [reply package/cloudshell/iac/terraform/services/sandbox_messages.py]: "in line 22 string formatting is not needed" -> Please explain
<span style="text-transform:lowercase"> -=> my attempt to overcome the uppercase css
<p style="display:inline"> -=> is needed cause if not a new line is created

File: package/cloudshell/iac/terraform/terraform_shell.py (hunk 7 above)
Comment (qualidan) [reply package/cloudshell/iac/terraform/terraform_shell.py]: unexplainable force...

File: package/cloudshell/iac/terraform/tagging/tag_terraform_resources.py [hunk 8]
```diff
@@ -576,7 +577,10 @@ def start_tagging_terraform_resources(main_dir_path: str, logger, tags_dict: dic
         # Error Code 3 mark to the outside tf script (that run me) that there was an error but not because of
         # the override procedure (but because the client tf file has compile errors even before we started the
         # override procedure)
-        exit(3)
+
+        # Had to change exit(3) to "raise" so exception can be handled outside
+        #exit(3)
+        raise ValueError("Exit before the override procedure began because the init/plan failed")
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/tagging/tag_terraform_resources.py]: I suggest the following error message:
"Error during Terraform Init/Plan when applying automated tags"

File: package/cloudshell/iac/terraform/tagging/tag_terraform_resources.py [hunk 9]
```diff
@@ -642,12 +646,16 @@ def start_tagging_terraform_resources(main_dir_path: str, logger, tags_dict: dic
                                          f" (Return_code is {return_code})"
                                          f"\n\nErrors are:\n{stderr}\n")
                 LoggerHelper.write_error("Tagging terraform resources operation has FAILED !!!!!")
-                exit(1)
+                # Had to change exit(1) to "raise" so exception can be handled outside
+                # exit(1)
+                raise ValueError("Errors were found in the last validation check")
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/tagging/tag_terraform_resources.py]: I suggest changing the message to:
"Validation errors during Terraform Init/Plan when applying automated tags"

Comment (alexazarh) [reply package/cloudshell/iac/terraform/tagging/tag_terraform_resources.py]: And its not a ValueError exception, maybe create a class like "TerraformAutoTagsError" for such exceptions?

File: package/cloudshell/iac/terraform/tagging/tag_terraform_resources.py [hunk 10]
```diff
@@ -642,12 +646,16 @@ def start_tagging_terraform_resources(main_dir_path: str, logger, tags_dict: dic
...
 
         else:
             LoggerHelper.write_warning("No untaggable resources were found, but errors in plan file do exists")
             LoggerHelper.write_error("Tagging terraform resources operation has FAILED !!!!!")
-            exit(1)
+            # Had to change exit(1) to "raise" so exception can be handled outside
+            # exit(1)
+            raise ValueError("No untaggable resources were found, but errors in plan file do exists")
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/tagging/tag_terraform_resources.py]: Suggested error message:
"No untaggable resources were found, but there is an error in Terraform Plan when applying automated tags. Please check the logs for more details"

File: package/cloudshell/iac/terraform/services/object_factory.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+from cloudshell.iac.terraform.tagging.tags import TagsManager
+
+
+class ObjectFactory:
+
+    @staticmethod
+    def create_tf_proc_executer(config: TerraformShellConfig,
+                                sandbox_data_handler: SandboxDataHandler,
+                                shell_helper: ShellHelperObject,
+                                tf_working_dir: str):
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/services/object_factory.py]: can you add return type?

**Review by alexazarh — CHANGES_REQUESTED** [review]

File: package/cloudshell/iac/terraform/tagging/tag_terraform_resources.py [hunk 12]
```diff
@@ -647,14 +649,15 @@ def start_tagging_terraform_resources(main_dir_path: str, logger, tags_dict: dic
                 LoggerHelper.write_error("Tagging terraform resources operation has FAILED !!!!!")
                 # Had to change exit(1) to "raise" so exception can be handled outside
                 # exit(1)
-                raise ValueError("Errors were found in the last validation check")
+                raise TerraformAutoTagsError("Errors were found in the last validation check")
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/tagging/tag_terraform_resources.py]: should be something like:
""Validation errors during Terraform Init/Plan when applying automated tags""

**Review by alexazarh — CHANGES_REQUESTED** [review]

**Review by alexazarh — APPROVED** [review]
